docs/source/kinds_of_types.rst

```diff
@@ -701,3 +701,207 @@ For more details, see :ref:`type-variable-value-restriction`.
 
    How ``bytes``, ``str``, and ``unicode`` are handled between Python 2 and
    Python 3 may change in future versions of mypy.
+
+.. _generators:
+
+Generators
+**********
+
+A generator can be annotated by the generic type
+``Generator[YieldType, SendType, ReturnType]``. For example:
+
+.. code-block:: python
+
+   def echo_round() -> Generator[int, float, str]:
+       sent = yield 0
+       while sent >= 0:
+           sent = yield round(sent)
+       return 'Done'
+
+Note that unlike many other generics in the typing module, the ``SendType`` of
+``Generator`` behaves contravariantly, not covariantly or invariantly.
+
+If your generator will only yield values, set the ``SendType`` and
+``ReturnType`` to None:
+
+.. code-block:: python
+
+   def infinite_stream(start: int) -> Generator[int, None, None]:
+       while True:
+           yield start
+           start += 1
+
+Alternatively, you may annotate your generator as having a return type of
+``Iterator[YieldType]``:
+
+.. code-block:: python
+
+   def infinite_stream(start: int) -> Iterator[int]:
+       while True:
+           yield start
+           start += 1
+
+.. _async-and-await:
+
+Typing async/await
+******************
+
+Mypy supports the ability to type coroutines that use the ``async/await``
+syntax introduced in Python 3.5. For more information regarding coroutines and
+this new syntax, see `PEP 492 <https://www.python.org/dev/peps/pep-0492/>`_.
+
+Functions that use the ``async`` keyword are typed just like normal functions.
+The return type annotation should be the same as the type of the value you
+expect to get back when ``await``-ing the coroutine.
+
+.. code-block:: python
+
+   import asyncio
+
+   async def format_string(tag: str, count: int) -> str:
+       return 'T-minus {} ({})'.format(count, tag)
+
+   async def countdown_1(tag: str, count: int) -> str:
+       while count > 0:
+           my_str = await format_string(tag, count)  # has type 'str'
+           print(my_str)
+           await asyncio.sleep(0.1)
+           count -= 1
+       return "Blastoff!"
+
+   loop = asyncio.get_event_loop()
+   loop.run_until_complete(countdown_1("Millennium Falcon", 5))
+   loop.close()
+
+The result of calling an ``async def`` function will be a value of type
+``Awaitable[T]``:
+
+.. code-block:: python
+
+   my_coroutine = countdown_1("Millennium Falcon", 5)
+   reveal_type(my_coroutine)  # has type 'Awaitable[str]'
+
+If you want to use coroutines in older versions of Python that do not support
+the ``async def`` syntax, you can instead use the ``@asyncio.coroutine``
+decorator to convert a generator into a coroutine. Note that we set the
+``YieldType`` to be ``Any`` since the functions we call ``yield from`` on may
+potentially return many different kinds of objects.
```

Comment:
Hm, the reason is actually more that yield type is an implementation detail shared between the coroutine runner (i.e. the asyncio event loop) and the Future implementation, and coroutines shouldn't have to know or care what type it actually is.